File: 2020/10/2020-10.jl

td1 = [16,10,15,5,1,11,7,19,6,12,4]
td2 = [ 28, 33, 18, 42, 31, 14, 46, 20, 48, 47, 24, 23, 49, 45, 19, 38, 39, 11, 1, 32, 25, 35, 8, 17, 7, 9, 4, 2, 34, 10, 3]

function jolt(data)
    js = sort(data)
    pushfirst!(js, 0)
    push!(js, js[end]+3)
    dfs = [js[i+1] - js[i] for i in 1:(length(js)-1)]
    cs = zeros(Int, 3)
    for df in dfs
        cs[df] += 1
    end
    return cs[1], cs[3]
end

function part1(data)
    n1,n3 = jolt(data)
    println("$n1 $n3 $(n1 * n3)")
    return n1 * n3
end

@assert part1(td1) == 35
@assert part1(td2) == 220

read_data(inp) = [parse(Int,s) for s in split(strip(inp), "\n")]
data = read_data(read("input.txt",String))
println("PART1: $(part1(data))")

# this approach blows up for large lists
function paths_recursive(js, i)
    if i == length(js)
        1
    else
        j = i + 1
        s = 0
        while j <= length(js) && js[j] - js[i] < 4
            s += paths_recursive(js, j)
            j += 1
        end
        s
    end
end

# efficient for large lists
function paths(js)
    ps = zeros(Int, length(js))
    ps[end] = 1
    for i = (length(js)-1):-1:1
        j = i + 1
        while j <= length(js) && (js[j] - js[i]) <= 3
            ps[i] += ps[j]
            j += 1
        end
    end
    ps[1]
end

function part2(data)
    js = sort(data)
    pushfirst!(js, 0)
    push!(js, js[end] + 3)
    #paths_recursive(js, 1)
    paths(js)
end

@assert part2(td1) == 8
@assert part2(td2) == 19208

println("PART2: ", part2(data))

File: 2020/10/2020-10.exs
ExUnit.start
defmodule Advent10.Test do
  use ExUnit.Case
  def test_data, do: "16 10 15 5 1 11 7 19 6 12 4"
  def test_data2, do: "28 33 18 42 31 14 46 20 48 47 24 23 49 45 19 38 39 11 1 32 25 35 8 17 7 9 4 2 34 10 3"

  def read_data(inp) do
    String.trim(inp)
    |> String.split(~r/[ \n]/)
    |> Enum.map(&String.to_integer(&1))
  end

  def part1(data) do
    xs = [0 | data] |> Enum.sort()
    cs = Enum.zip(xs, (tl xs))
         |> Enum.map(fn {a,b} -> b - a end)
         |> Enum.frequencies()
    cs[1] * (cs[3] + 1)
  end

  def part2(data) do
    xs = [0 | data] |> Enum.sort() |> Enum.reverse()
    mx = hd xs
    {_n,s} = Enum.reduce(xs, [{mx+3, 1}],
      fn x,ls -> 
        #IO.puts "trying #{x} #{inspect ls}"
        sm = ls
        |> Enum.take_while(fn {n,_s} -> n - x <= 3 end)
        |> Enum.reduce(0, fn {_n,s},acc -> acc + s end)
        [{x, sm} | ls]
      end) |> hd()
    s
  end
    
  test "test" do
    td = read_data(test_data())
    td2 = read_data(test_data2())
    assert part1(td) == 35
    assert part1(td2) == 220
    data = read_data(File.read!("input.txt"))
    IO.puts part1(data)

    assert part2(td) == 8
    assert part2(td2) == 19208
    IO.puts "PART2: #{part2(data)}"
  end
end
